runtime/src/accounts.rs

```diff
@@ -320,36 +332,79 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            let credit_only_locks = credit_only_locks.read().unwrap();
+            if locks.contains(k)
+                || (credit_only_locks.contains_key(k)
+                    && *credit_only_locks
+                        .get(&k)
+                        .unwrap()
+                        .lock_count
+                        .lock()
+                        .unwrap()
+                        > 0)
+            {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
```

Comment:
Simliar to the comment above, this check needs to happen while the `lock_count.lock`, is held, otherwise, a credit-debit transaction could come in after this lock is released

Reply:
This one I disagree with, actually. The credit-debit lock HashSet is locked throughout this entire method, so there isn't any way a credit-debit lock could sneak in between this check and the credit-only locking below.

Reply:
oh i see, that lock must be held for the entire duration of this function. That should work!